home_audit.ts

```diff
@@ -0,0 +1,691 @@
+        const {heating, cooling} = system;
+        heating && hvac_types.push(heating.type);
+        cooling && hvac_types.push(cooling.type);
+    });
+    if(!hvac_types.includes('boiler') && category === 'combined') {
+        addErrorMessage(`systems/domestic_hot_water/category`, 'Must have a boiler for combined hot water category');
+    }
+}
+
+/**
+ * Check that fuel is appropriate for the hot water system
+ */
+function checkHotWaterFuelValid(hot_water) {
+    const {type, fuel_primary} = hot_water;
+    if(['tankless_coil', 'indirect'].includes(type) && !nullOrUndefined.includes(fuel_primary)) {
+        addErrorMessage(`systems/domestic_hot_water/fuel_primary`, 'Fuel is only used if type is set to storage or heat pump');
+    } else if(type === 'heat_pump' && fuel_primary !== 'electric') {
+        addErrorMessage(`systems/domestic_hot_water/fuel_primary`, 'Fuel must be electric if type is heat pump');
+    }
+}
+
+/**
+ * Check that efficiency is appropriate for the hot water system
+ */
+function checkHotWaterEfficiencyValid(hot_water) {
+    const {type, efficiency_method} = hot_water;
+    if(['heat_pump', 'tankless', 'tankless_coil'].includes(type) && efficiency_method === 'shipment_weighted') {
+        addErrorMessage(`systems/domestic_hot_water/efficiency_method`, 'Invalid Efficiency Method for entered Hot Water Type');
+    }
+}
+
+/**
+ * Check that the year is appropriate for the hot water system
+ */
+function checkHotWaterYearValid(hot_water) {
+    const {year} = hot_water;
+    if(year && (1972 > year || (new Date()).getFullYear() < year)) {
+        addErrorMessage(`systems/domestic_hot_water/year`, `Invalid year, must be between 1972 and ${(new Date()).getFullYear()}`)
+    }
+}
+
+/**
+ * Check that the energy factor is valid for the hot water system
+ */
+function checkHotWaterEnergyFactorValid(hot_water) {
+    const {type, energy_factor} = hot_water;
+    if(["indirect", "tankless_coil"].includes(type) && !nullOrUndefined.includes(energy_factor)) {
+        addErrorMessage(`systems/domestic_hot_water/energy_factor`, `Energy Factor not valid for selected Hot Water Type`);
+    }
+    let min,max;
+    if (type === 'storage') {
+        [min, max] = [0.45, 0.95];
+    } else if (type === 'tankless') {
+        [min, max] = [0.45, 0.99];
+    } else if (type === 'heat_pump') {
+        [min, max] = [1, 4];
+    }
+    if(energy_factor && (energy_factor < min || energy_factor > max)) {
+        addErrorMessage(`systems/domestic_hot_water/energy_factor`, `${energy_factor} is outside the allowed range (${min} - ${max})`);
+    }
+}
+
+/**
+ * Check that the solar electric system is of a valid year
+ */
+function checkSolarElectricYearValid(solar_electric) {
+    const {year} = solar_electric;
+    if(year && (2000 > year || (new Date()).getFullYear() < year)) {
+        addErrorMessage(`systems/generation/solar_electric/year`, `Invalid year, must be between 2000 and ${(new Date()).getFullYear()}`)
+    }
+}
```

Comment:
I am excited for the TS refactor and all of these being typed :)

Reply:
I'm working on it! It's going to be lovely :)